### PR TITLE
E2e android

### PR DESCRIPTION
**Pull request by bestander**
# End to End testing on Android emulator for CI

- converted shell script `scripts/e2e-test.sh` into JS script to have more programming flexibility
- using appium execute 2 tests after a fresh React Native app installation: check HMR and that debugging mode does not crash the app
- made sure tests can be stable on limited CI systems and added ways to debug any problems in the future

Using appium we can now interact with Android app and test its state.
As a follow up i am planning to write a blog post on how to use appium with android and ios for e2e testing.

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -71,13 +71,12 @@ test:
     # run installed apk with tests
     - source scripts/circle-ci-android-setup.sh && retry3 ./scripts/run-android-instrumentation-tests.sh com.facebook.react.tests
 
-    # Deprecated: run tests with Gradle, we keep them for a while to compare performance
     - ./gradlew :ReactAndroid:testDebugUnitTest
-    - ./gradlew :ReactAndroid:installDebugAndroidTest
-    - source scripts/circle-ci-android-setup.sh && retry3 ./scripts/run-android-instrumentation-tests.sh com.facebook.react.tests.gradle
+    # Deprecated: these tests are executed using Buck above, while we support Gradle we just make sure the test code compiles
+    - ./gradlew :ReactAndroid:assembleDebugAndroidTest
```
Comment (mkonicek) [inline circle.yml]: Would Buck fails if the tests don't compile? Can you remove this step completely now?

File: local-cli/generator-android/templates/src/app/BUCK [hunk 2]
```diff
@@ -5,7 +5,7 @@ import re
 # - install Buck
 # - `npm start` - to start the packager
 # - `cd android`
-# - `cp ~/.android/debug.keystore keystores/debug.keystore`
+# - `keytool -genkey -v -keystore keystores/debug.keystore -storepass android -alias androiddebugkey -keypass android -dname "CN=Android Debug,O=Android,C=US`
```
Comment (mkonicek) [inline local-cli/generator-android/templates/src/app/BUCK]: 👍

File: scripts/android-e2e-test.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
+'use strict';
+
+/**
+ * Used in run-ci-e2e-test.js and executed in Travis and Circle CI.
+ * E2e test that verifies that init app can be installed, compiled, started and Hot Module Reloading with Chrome Debugging work.
```
Comment (mkonicek) [inline scripts/android-e2e-test.js]: nit: Hot Module reloading and Chrome debugging work

File: scripts/android-e2e-test.js [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+  before(function () {
+    driver = wd.promiseChainRemote({
+      host: 'localhost',
+      port: 4723
+    });
+    driver.on('status', function (info) {
+      console.log(info.cyan);
+    });
+    driver.on('command', function (meth, path, data) {
```
Comment (mkonicek) [inline scripts/android-e2e-test.js]: Why is this called `meth`?

Comment (bestander) [inline scripts/android-e2e-test.js]: lol, that is what they called it in appium.
I'll give it a better name

File: .travis.yml [hunk 5]
```diff
@@ -16,16 +16,15 @@ install:
 
 script:
   - if [[ "$TEST_TYPE" = objc ]]; then travis_retry ./scripts/objc-test.sh; fi
-  - if [[ "$TEST_TYPE" = e2e-objc ]]; then travis_retry ./scripts/e2e-test.sh --ios; fi
+  - if [[ "$TEST_TYPE" = e2e-objc ]]; then travis_retry node ./scripts/run-ci-e2e-tests.js --ios; fi
   - if [[ "$TEST_TYPE" = js ]]; then npm run flow check; fi
   - if [[ "$TEST_TYPE" = js ]]; then npm test -- --maxWorkers=1; fi
-  - if [[ "$TEST_TYPE" = js ]]; then travis_retry ./scripts/e2e-test.sh --packager; fi
```
Comment (mkonicek) [inline .travis.yml]: Did this run the packager test with npm2? Why remove it?

File: circle.yml [hunk 6]
```diff
@@ -71,13 +71,12 @@ test:
...
 
-    # JS and Android e2e test
-    - source scripts/circle-ci-android-setup.sh && retry3 ./scripts/e2e-test.sh --packager --android
```
Comment (mkonicek) [inline circle.yml]: Why kill the packager tests? Because the Android test already covers it?

File: scripts/run-ci-e2e-tests.js [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+  const packagerProcess = spawn('npm', ['start'], 
+  {
+    stdio: 'inherit',
+    env: packagerEnv
+  });
+  SERVER_PID = packagerProcess.pid;
+  echo(`Starting packager server, ${SERVER_PID}`);
+  exec('sleep 5s');
+  // prepare cache to reduce chances of possible red screen "Can't fibd variable __fbBatchedBridge..."
+  exec('response=$(curl --write-out %{http_code} --silent --output /dev/null localhost:8081/index.ios.bundle?platform=ios)');
```
Comment (mkonicek) [inline scripts/run-ci-e2e-tests.js]: Why is this needed in the iOS test but not Android?

File: scripts/run-ci-e2e-tests.js [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    exit(cleanup(1));
+  }
+}
+
+if (args.indexOf('--ios') !== -1) {
+  echo('Running an iOS app');
+  cd('ios');
+  // Make sure we installed local version of react-native
+  if (!test('-e', path.join('EndToEndTest', path.basename(MARKER_IOS)))) {
+    echo('iOS marker was not found, react native init command failed?');
```
Comment (mkonicek) [inline scripts/run-ci-e2e-tests.js]: nit: react-native init

File: .travis.yml (hunk 5 above)
Comment (bestander) [inline .travis.yml]: My point was that packager is tested in --ios mode anyway.
But I've had a better look, it was actually testing flow as well.
I returned this test to both IOS/Android and named it `--js` to emphasize that we are testing js related stuff

File: circle.yml (hunk 6 above)
Comment (bestander) [inline circle.yml]: fixed

File: scripts/run-ci-e2e-tests.js (hunk 7 above)
Comment (bestander) [inline scripts/run-ci-e2e-tests.js]: Android e2e test is smarter.
If it sees a red screen it tries to reload the app and then it works.
Both solutions are a hack that indicate a race condition we have in the framework.

File: scripts/run-ci-e2e-tests.js (hunk 8 above)
Comment (bestander) [inline scripts/run-ci-e2e-tests.js]: fixed

**Comment by mkonicek** [comment]
Amazing work Konstantin!

@facebook-github-bot shipit
